bpo-19382: Adding test cases for module tabnanny.

ultimatecoder opened the pull request:
* **Description:** Unit tests for standard module `tabnanny`. Tests are added for almost all the functionality except `tabnanny.Whitespace` class.

* **Reason leaving Whitespace:?** I found the module contains mathematical calculations. I will try to understand it first and then write tests for it. I think it will be good to do that in another branch. 

* **Testing strategy:** Whitebox

* **BPOs**
 
    * https://bugs.python.org/issue19382
    * https://bugs.python.org/issue19102

<!-- issue-number: bpo-19382 -->
https://bugs.python.org/issue19382
<!-- /issue-number -->

ultimatecoder committed:
[ADD] Adding test cases for module tabnanny. Testing stratergy:
Whitebox.

DimitrisJim commented:
Hi @ultimatecoder, you'll also need to remove `tabnanny` from the [`untested` tuple in `test_sundry`](https://github.com/python/cpython/blob/master/Lib/test/test_sundry.py#L9). This is one of the failures AppVeyor (and soon Travis, I'd guess), reports.

vadmium commented on Lib/test/test_tabnanny.py:
```diff
@@ -0,0 +1,353 @@
+
+        Use this method to assert expected values of `stdout` and `stderr` after
+        running tabsize.check() on given `dir` or `file` path. Because
+        tabsize.check() captures arised exceptions and writes to `stdout` and
```
Just write “captures exceptions”  (or change arised → raised).

What is “tabsize”? Did you mean “tabnanny.check”?

vadmium commented on Lib/test/test_tabnanny.py:
```diff
@@ -0,0 +1,353 @@
+                                     Returns path of the created file. Returned
+                                     file path can be feeded to
+                                     `tabnanny.check()`.  Here, the expression
+                                     `*` at function name is dipicting the
```
feeded → fed
dipicting → depicting

vadmium commented on Lib/test/test_tabnanny.py:
```diff
@@ -0,0 +1,353 @@
+            * dir_or_file : A path to file or directory. This argument will be
+                            feeded directly to tabnanny.check().
+            * out         : Expacted `stdout` after running tabsize.check().
+            * err         : Expacted `stderr` after running tabsize.check().
```
feeded → fed
Expacted → Expected

vadmium commented on Lib/test/test_tabnanny.py:
```diff
@@ -0,0 +1,353 @@
+        """Directory containing few error free python source code files.
+
+        Because order of files returned by `os.lsdir()` is not fixed, verifying
+        existance of each output lines at `stdout` using `in` operator.
```
verifying → verify the
existance → existence

vadmium commented on Lib/test/test_tabnanny.py:
```diff
@@ -0,0 +1,353 @@
+        self._verify_tabnanny_check(file_path)
+
+    def test_correct_directory_verbose(self):
+        """Directory containing few error free python source code files.
```
few → a few [otherwise, the implication is that most of the files have errors: few are error-free]

vadmium commented on Lib/test/test_tabnanny.py:
```diff
@@ -0,0 +1,353 @@
+
+    def test_correct_file(self):
+        """A python source code file without any errors."""
+        file_path = create_python_file(SOURCE_CODES["error_free"])
```
It seems these files are never removed

vadmium commented on Lib/test/test_tabnanny.py:
```diff
@@ -0,0 +1,353 @@
+
+        Because order of files returned by `os.lsdir()` is not fixed, verifying
+        existance of each output lines at `stdout` using `in` operator.
+        `verbose` mode of `tabnanny.verbose` asserts `stdout`.
```
This line does not make sense. Is it needed? It seems fairly obvious from the method name and code that you are exercising the tabnanny.verbose flag.

vadmium commented on Lib/test/test_tabnanny.py:
```diff
@@ -0,0 +1,353 @@
+    def test_when_tokenize_tokenerror(self):
+        """A python source code file eligible for raising 'tokenize.TokenError'.
+        """
+        file_path = create_python_file(SOURCE_CODES["incomplete_expression"])
```
Some of these test code strings are specific to one test case. IMO they would be easier to read if moved closer to the relevant test methods. They do not need to be in a shared module-level array.

vadmium commented on Lib/test/test_tabnanny.py:
```diff
@@ -0,0 +1,353 @@
+
+        # "tab_space_errored_1": exicutes block under type == tokenize.INDENT
+        #                        at `tabnanny.process_tokens()`.
+        # "tab space_errored_2": exicutes block under
```
exicutes → executes

vadmium commented on Lib/test/test_tabnanny.py:
```diff
@@ -0,0 +1,353 @@
+        self._verify_tabnanny_check(file_path, out=msg)
+
+    def test_when_no_file(self):
+        """A python file which does not exists actually in system."""
```
exists → exist

vadmium commented on Lib/test/test_tabnanny.py:
```diff
@@ -0,0 +1,353 @@
+    def test_when_no_file(self):
+        """A python file which does not exists actually in system."""
+        file_path = 'no_file.py'
+        msg = "%r: I/O Error: [Errno 2] No such file or directory: %r\n" % (
```
Doesn’t the message come from the OS (strerror)? Does it work on Windows, for example?

ultimatecoder replied:
I am confused for this. There are code samples which are used as input to more than one class. Because of that reason, I believe it is good to keep one module level non-repeated copy of them. It will be more effective if we want to update particular situation of copy of code.

vadmium replied:
And I am confused by your reply too! :) Let me start again:

The test is checking the exact error message (“No such file or directory”) and errno value (2). I suspect these come from the operating system. Maybe they vary on different platforms, user languages, etc. If so, the tests shouldn’t rely on these values always being the same. It may be better to rewrite the test to just check that an error message is printed, without checking too many details of the message contents.

vadmium commented on Lib/test/test_tabnanny.py:
```diff
@@ -0,0 +1,353 @@
+    """Testing `tabnanny.format_witnesses()`."""
+
+    def test_format_whitenesses(self):
+        """Asserting formatter result by givinig various input samples."""
```
giving

vadmium commented on Lib/test/test_tabnanny.py:
```diff
@@ -0,0 +1,353 @@
+        self._validate_cmd(stderr=stderr)
+
+    def test_quiet_flag(self):
+        """Should not display less when quite mode is on."""
```
quiet mode?
“less” looks wrong, unless you meant “Should display less”

vadmium commented on Lib/test/test_tabnanny.py:
```diff
@@ -0,0 +1,353 @@
+    def test_quiet_flag(self):
+        """Should not display less when quite mode is on."""
+        file_path = create_python_file(SOURCE_CODES["nannynag_errored"])
+        stdout = b"%s\n" % (file_path.encode('ascii'),)
```
Don’t you get CRLF on Windows?

vadmium commented on Lib/test/test_tabnanny.py:
```diff
@@ -0,0 +1,353 @@
+    def test_with_errored_file(self):
+        """Should displays error when errored python file is given."""
+        file_path = create_python_file(SOURCE_CODES["wrong_indented"])
+        stderr = b"%r: Indentation Error: " % (file_path,)
```
In Python 3, b"%r" is equivalent to calling the “ascii” function. I expect the test will fail if the file path has non-ASCII characters. It may be better to only test that an error is being reported, but tolerate variations in the message.

vadmium commented on Lib/test/test_tabnanny.py:
```diff
@@ -0,0 +1,353 @@
+    def test_command_usage(self):
+        """Should display usage on no arguments."""
+        path = findfile('tabnanny.py')
+        stderr = b'Usage: %s [-v] file_or_directory ...' % (path.encode('ascii'))
```
What if the path is non-ASCII? Again, just check the general form of the output and don’t be too specific.

ultimatecoder replied:
I don't think this https://github.com/python/cpython/blob/master/Lib/tempfile.py#L144 generates anything Non-ASCII.

vadmium replied:
No, but the OS can provide a parent directory for temporary files that has non-ASCII path components. It seems relatively common on Windows that the temporary directory is within a user’s profile, and the profile directory can be a localized non-ASCII name. Top Google result: <https://bugzilla.mozilla.org/show_bug.cgi?id=260034>.

ultimatecoder replied:
Many thanks for affording time for reviewing this. I will agree with you on this point. I have updated code. Will you please review again?

vadmium commented:
Check out the existing patches and reviews you linked. You may be repeating some of the same problems, e.g. CRLF on Windows.

ultimatecoder commented:
@vadmium Many thanks for your comments. I am not native English speaker. I have fixed the mistakes. I have added comments to some comments. Please review my fixes. Thanks!

@DimitrisJim Thanks for your comment. I have removed `tabnanny` from untested modules list. Thanks!

vadmium commented:
Sorry but I don’t have an easy way to see your fixes relative to the old version I reviewed. I’m not really set up to properly review and push stuff with the new Git Hub setup, so I will leave this to someone else, or for another time in the future.

vstinner commented on Lib/test/test_tabnanny.py:
```diff
@@ -0,0 +1,348 @@
+"""Testing `tabnanny` module.
+
```
remove one empty line

ultimatecoder replied:
Removed this. This is Github glitch.

vstinner commented on Lib/test/test_tabnanny.py:
```diff
@@ -0,0 +1,348 @@
+    * errored    : Whitespace related problems present in file.
+"""
+from unittest import TestCase, mock
+from unittest.mock import patch
```
Hum, I prefer to write mock.patch, please remove this import

vstinner commented on Lib/test/test_tabnanny.py:
```diff
@@ -0,0 +1,348 @@
+from unittest.mock import patch
+import tabnanny
+import tokenize
+from tempfile import NamedTemporaryFile, TemporaryDirectory
```
Please import tempfile and write tempfile.NamedTemporaryFile.

vstinner commented on Lib/test/test_tabnanny.py:
```diff
@@ -0,0 +1,348 @@
+
+    It creates file with `.py` extension.
+
+    Returns: Absolute path of created code file."""
```
IMHO this docstring is useless. Usually, we don't comment much tests, they are not intented to end users.

vstinner commented on Lib/test/test_tabnanny.py:
```diff
@@ -0,0 +1,348 @@
+
+    Returns: Absolute path of created code file."""
+    with NamedTemporaryFile(mode='w', dir=dir,
+                            delete=False, suffix=".py") as f:
```
If you only care of creating a temporary filename, use tempfile.mktemp() and then open() manually the file to create it. You can use open(tmpname, "x") to prevent race conditions.

vstinner commented on Lib/test/test_tabnanny.py:
```diff
@@ -0,0 +1,348 @@
+    def test_errprint(self):
+        """Asserting result of `tabnanny.errprint()` by giving sample inputs."""
+        tests = [
+            {'args': ['first', 'second'], 'expected': 'first second\n'},
```
Instead of a dict, you can simply use a tuple (args, expected)

ultimatecoder replied:
Don't think we should kill the readability by converting it to `tuple`. How about `collections.namedtuple`?

ultimatecoder replied:
Or How about this way?
```
        tests = [
         #  (Function arguments , Expected output)
            (['first', 'second'], 'first second\n'),
            (['first']          , 'first\n'),
            ([1, 2, 3]          , '1 2 3\n'),
            ([]                 , '\n')
        ]
```

merwok replied:
Keep tests simple:
```python
for args, expected in [
    ([1], '1'),
  ]:
```

brettcannon replied:
Yes, keep them simple. And don't use namedtuples except for keeping backwards-compatibility with an interface that used to take an indexed object and now takes an object with attributes.

vstinner commented on Lib/test/test_tabnanny.py:
```diff
@@ -0,0 +1,348 @@
+
+
+class TestNannyNag(TestCase):
+    """Testing `tabnanny.NannyNag` exception."""
```
Not sure that this docstring is super useful.

vstinner commented on Lib/test/test_tabnanny.py:
```diff
@@ -0,0 +1,348 @@
+            self.assertEqual(stderr.getvalue() , test['expected'])
+
+
+class TestNannyNag(TestCase):
```
I don't think that it's worth it to declare one test case to test a single class/function. Try to put most or all tests into a single test case called "TabnannyTestCase".

TestCheck deserves its own test case since it has a setUp() method.

vstinner commented on Lib/test/test_tabnanny.py:
```diff
@@ -0,0 +1,348 @@
+class TestNannyNag(TestCase):
+    """Testing `tabnanny.NannyNag` exception."""
+    def setUp(self):
+        self.tests = [
```
Please move this list into the method directly.

vstinner commented on Lib/test/test_tabnanny.py:
```diff
@@ -0,0 +1,348 @@
+    """Testing tabnanny.check()."""
+
+    def setUp(self):
+        tabnanny.verbose = 0  # Forcefully deactivating verbose mode.
```
Unit tests must have zero side effect: you must save the old verbose value and restore it. Example:

```
self.addCleanup(setattr, tabnanny, 'verbose', tabnanny.verbose)
tabnanny.verbose = 0
```

ultimatecoder replied:
Aha! a geek way :+1:

vstinner commented on Lib/test/test_tabnanny.py:
```diff
@@ -0,0 +1,348 @@
+    def setUp(self):
+        tabnanny.verbose = 0  # Forcefully deactivating verbose mode.
+
+    def _verify_tabnanny_check(self, dir_or_file, out="", err=""):
```
No need to make this method private, remove "_" prefix.

vstinner commented on Lib/test/test_tabnanny.py:
```diff
@@ -0,0 +1,348 @@
+            * dir_or_file : A path to file or directory. This argument will be
+                            fed directly to tabnanny.check().
+            * out         : Expected `stdout` after running tabnanny.check().
+            * err         : Expected `stderr` after running tabnanny.check().
```
Again, I'm not sure that a long comment is needed for a simple unit test.

ultimatecoder replied:
I think it is explaining why the method is here. Suggest a way for improving the docstring.

brettcannon replied:
At least ditch the whole `Arguments` section; we don't use that style in the stdlib.

vstinner commented on Lib/test/test_tabnanny.py:
```diff
@@ -0,0 +1,348 @@
+
+    def test_correct_file(self):
+        """A python source code file without any errors."""
+        file_path = create_tmp_python_file(SOURCE_CODES["error_free"])
```
Who removes the temporary file? I suggest to put create_tmp_python_file() in the test case (or in a base class if it's needed by multiple test cases) and use self.addCleanup(support.unlink, tmpfile) to make sure that the temporary file is removed.

ultimatecoder replied:
Adding it to a method will be a poor idea. If I will add it to base class then we will have a base class with one method. Will you accept that?

brettcannon replied:
Yes, if the method needs to access information that will be on the instance then a class with a single method is fine. But if you aren't reusing it then just inline the function.

vstinner commented on Lib/test/test_tabnanny.py:
```diff
@@ -0,0 +1,348 @@
+
+    @patch('tabnanny.NannyNag')
+    def test_with_correct_code(self, MockNannyNag):
+        """A python source code without any whitespace related problems."""
```
IMHO the docstring is useless.

vstinner commented on Lib/test/test_tabnanny.py:
```diff
@@ -0,0 +1,348 @@
+            exception = tabnanny.NannyNag
+            with open(file_path) as f:
+                tokens = tokenize.generate_tokens(f.readline)
+                self.assertRaises(exception, tabnanny.process_tokens, tokens)
```
Move this assert out of the with block.

ultimatecoder replied:
We can't move this out.

brettcannon replied:
Why not? Is `tokenize.generate_tokens()` lazy (which isn't documented)?

ultimatecoder replied:
@brettcannon The `tokenize.generate_tokens()` is a lazy function (generator). Can you help to identify a way which allows putting `assertRaises` outside the `with`? Thanks!

vstinner commented on Lib/test/test_tabnanny.py:
```diff
@@ -0,0 +1,348 @@
+        """
+        rc, out, err = script_helper.assert_python_ok('-m', 'tabnanny', *args)
+        # Note: The `splitlines()` will solve the problem of CRLF(\r) added
+        # by OS Windows.
```
Use universal_newlines=True to normalize newlines. I also prefer to use text rather than bytes for stdout/stderr.

ultimatecoder replied:
I don't think these chain of methods, `assert_python_ok` -> `_assert_python` -> `run_python_until_end` is allowing to pass `universal_newlines` keyword argument!

Reference: https://github.com/python/cpython/blob/master/Lib/test/support/script_helper.py#L94

Are you suggesting to replace `script_helper.assert_python_ok()` with `subprocess.Popen()` here?

vstinner commented on Lib/test/test_tabnanny.py:
```diff
@@ -0,0 +1,348 @@
+        # Note: The `splitlines()` will solve the problem of CRLF(\r) added
+        # by OS Windows.
+        self.assertListEqual(out.splitlines(), stdout.splitlines())
+        self.assertEqual(rc, return_code)
```
assert_python_ok() already tests that returncode==0.

vstinner commented on Lib/test/test_tabnanny.py:
```diff
@@ -0,0 +1,348 @@
+class TestCommandLine(TestCase):
+    """Tests command line interface of `tabnanny`."""
+
+    def _validate_cmd(self, *args, return_code=0, stdout=b"", stderr=b""):
```
make this method public

vstinner commented on Lib/test/test_tabnanny.py:
```diff
@@ -0,0 +1,348 @@
+    def test_double_verbose_mode(self):
+        """Should display detailed error information if double verbose is on."""
+        file_path = create_tmp_python_file(SOURCE_CODES["nannynag_errored"])
+        stdout = b"checking %r ...\n" % (file_path,)
```
For better readability, you can use textwrap.dedent() and a multiline """...""" indented string. See test_faulthandler for examples.

ultimatecoder replied:
This is a nice suggestion :+1: Are you expecting these multiline strings to convert into a regular expression? I think this way just looks fine.

vstinner reviewed (changes requested):
I like the overall change, but I have a long list of requests :-)

JelleZijlstra commented on Lib/test/test_tabnanny.py:
```diff
@@ -0,0 +1,347 @@
+class TestFormatWitnesses(TestCase):
+    """Testing `tabnanny.format_witnesses()`."""
+
+    def test_format_whitenesses(self):
```
witnesses, not whitenesses

brettcannon commented on Lib/test/test_tabnanny.py:
```diff
@@ -0,0 +1,356 @@
+            ([]                 , '\n')
+        ]
+
+        for args, expected in tests:
```
I would use subtests here: https://docs.python.org/3/library/unittest.html#distinguishing-test-iterations-using-subtests

vstinner replied:
Done

brettcannon commented on Lib/test/test_tabnanny.py:
```diff
@@ -0,0 +1,356 @@
+        `verbose` mode of `tabnanny.verbose` asserts `stdout`.
+        """
+        with tempfile.TemporaryDirectory() as tmp_dir:
+            lines = ["%r: listing directory\n" % (tmp_dir,),]
```
Use f-strings everywhere. 😄

vstinner replied:
Done

brettcannon commented on Lib/test/test_tabnanny.py:
```diff
@@ -0,0 +1,356 @@
+        """
+        with tempfile.TemporaryDirectory() as tmp_dir:
+            lines = ["%r: listing directory\n" % (tmp_dir,),]
+            with TemporaryPyFile(
```
To make this easier to read, I would create the context managers outside the `with` statement:
```python
file1 = TemporaryPyFile(...)
file2 = TemporaryPyFile(...)
with file1 as file1_path, file2 as file2_path:
    ...
```

vstinner replied:
For this specific case, IMHO "file1 = ..." is better.

brettcannon commented on Lib/test/test_tabnanny.py:
```diff
@@ -0,0 +1,356 @@
+    def test_correct_directory(self):
+        """Directory which contains few error free python source code files."""
+        with tempfile.TemporaryDirectory() as tmp_dir:
+            with TemporaryPyFile(
```
Move the `TemporaryPyFile` instantiation to its own line so you don't have to spread over multiple lines. (Same of the other lines where you were forced to spread a `with` statement over multiple lines.)

vstinner replied:
I looked at the new code and it's fine.

brettcannon commented on Lib/test/test_tabnanny.py:
```diff
@@ -0,0 +1,356 @@
+
+    def test_when_tokenize_tokenerror(self):
+        """A python source code file eligible for raising 'tokenize.TokenError'.
+        """
```
Pull this up to the next line up (it's okay for  it to spill over 80 columns).

brettcannon commented on Lib/test/test_tabnanny.py:
```diff
@@ -0,0 +1,356 @@
+            SOURCE_CODES["nannynag_errored"]
+        ) as file_path:
+            out = "%r: *** Line 3: trouble in tab city! ***\n" % (file_path,)
+            out += "offending line: '\\tprint(\"world\")\\n'\n"
```
Don't worry too much about the exact output being formatted as it is. For instance, if someone tweaked the grammar of the sentence the test should still pass. Do still check for semantically important info, though, like that the offending line is in the output.

vstinner replied:
Maybe the test is too strict, but IMHO we have been too pendantic on this nice PR which has been proposed 3 months ago. tabnanny has currently no test. If the test fails tomorrow, we will just fix the test. No need to write the perfect test here.

brettcannon commented on Lib/test/test_tabnanny.py:
```diff
@@ -0,0 +1,356 @@
+    def test_when_no_file(self):
+        """A python file which does not exist actually in system."""
+        file_path = 'no_file.py'
+        err = "%r: I/O Error: [Errno 2] No such file or directory: %r\n" % (
```
You don't want to tie yourself to exception messages as they do change between versions. Just care about the right exception being raised.

ultimatecoder replied:
I agree, but the code does not raise an exception reference: https://github.com/python/cpython/blob/master/Lib/tabnanny.py#L99
What way will be better to assert the working of mentioned function?

brettcannon commented on Lib/test/test_tabnanny.py:
```diff
@@ -0,0 +1,356 @@
+        #                        `check_equal and type not in JUNK` condition at
+        #                        `tabnanny.process_tokens()`.
+
+        for key in ["tab_space_errored_1", "tab_space_errored_2"]:
```
Sub tests would be great here.

vstinner replied:
Done

brettcannon reviewed (dismissed):
Nothing fundamentally wrong, just some things to help make the tests easier to manage in future years.

vstinner reviewed (commented):
Please remove Lib/test/.test_tabnanny.py.swo from your PR.

vstinner commented on Lib/test/test_tabnanny.py:
```diff
@@ -0,0 +1,344 @@
+        """Asserting formatter result by giving various input samples."""
+        tests = [
+            ('Test' , 'at tab sizes T, e, s, t'),
+            (''     , 'at tab size '),
```
Sorry but this style is not compliant with PEP 8, please strip spaces before ",".

vstinner commented on Lib/test/test_tabnanny.py:
```diff
@@ -0,0 +1,344 @@
+
+    def __exit__(self, exc_type, exc_value, exc_traceback):
+        if exc_type is None:
+            unlink(self.file_path)
```
You should *always* remove the created file.

vstinner commented on Lib/test/test_tabnanny.py:
```diff
@@ -0,0 +1,344 @@
+
+    def validate_cmd(self, *args, stdout="", stderr=""):
+        """Common function to assert the behaviour of command line interface."""
+        _, out, err = script_helper.assert_python_ok('-m', 'tabnanny', *args)
```
You may use:

proc = script_helper.spawn_python('-m', 'tabnanny', *args, text=True)
out, err = proc.communicate()
self.assertEqual(out, stdout)
self.assertEqual(err, stderr)
self.assertEqual(proc.returncode, 0)

text=True normalizes newlines to \n (Unix EOL).

vstinner reviewed (commented):
Oh, I am sorry, I forgot about your old PR :-( Here is a new review. Your PR is very close to be ready to be merged. Sorry again about the slow review :-(

vstinner commented on Lib/test/test_tabnanny.py:
```diff
@@ -0,0 +1,344 @@
+
+    def __exit__(self, exc_type, exc_value, exc_traceback):
+        unlink(self.file_path)
+        return exc_type is None
```
I'm not sure that returning True is correct here. I just suggest to remove the "return" line.

ultimatecoder replied:
I agree, returning True isn't making sense here. Thanks for commenting.
